base/MAPL_ObsUtil.F90

```diff
@@ -375,33 +409,34 @@ function get_filename_from_template_use_index (obsfile_start_time, obsfile_inter
     call ESMF_time_to_two_integer(time, itime, _RC)
     nymd = itime(1)
     nhms = itime(2)
+    
+    ! parse time info
+    !
+    j= index(file_template, '*')
+    if (j>0) then
+       ! wild char exist
+       !!print*, 'pos of * in template =', j
+       file_template_left = file_template(1:j-1)
+       call fill_grads_template ( filename_left, file_template_left, &
+            experiment_id='', nymd=nymd, nhms=nhms, _RC )
+       filename= trim(filename_left)//trim(file_template(j:))
+    else
+       ! exact file name
+       call fill_grads_template ( filename, file_template, &
+            experiment_id='', nymd=nymd, nhms=nhms, _RC )
+    end if
 
-
-       j= index(file_template, '*')
-       if (j>0) then
-          ! wild char exist
-          !!print*, 'pos of * in template =', j
-          file_template_left = file_template(1:j-1)
-          call fill_grads_template ( filename_left, file_template_left, &
-               experiment_id='', nymd=nymd, nhms=nhms, _RC )
-          filename= trim(filename_left)//trim(file_template(j:))
-          cmd="bash -c 'ls "//trim(filename)//"' &> zzz_MAPL"
-          CALL execute_command_line(trim(cmd))
-          open(newunit=u, file='zzz_MAPL', status='unknown')
-          read(u, '(a)') filename
-          i=index(trim(filename), 'ls')
-          if (i==1) then
-             filename=''
-          end if
-          !       cmd="rm -f ./zzz_MAPL"
-          !       CALL execute_command_line(trim(cmd))
-          close(u)
-       else
-          ! exact file name
-          call fill_grads_template ( filename, file_template, &
-               experiment_id='', nymd=nymd, nhms=nhms, _RC )
-       end if
-
+    ! test on bash
+    !
+    cmd="bash -c 'ls "//trim(filename)//"' &> zzz_MAPL"
```

Comment:
Probably wrap in a helper function as well.   Think about the functionality - then we can at least have a test and think about wrapping C instead.

Reply:
@bena-nasa Could you look at this bit of code?   Is there already something in the ExtData utilities that might accomplish this?

Reply:
Thank you, @mathomp4  and @tclune, for getting to this point.  I had some discussion with @bena-nasa about this the week before and he can give more information.

The code tries to get file names from  MOD04_L2.A2017090.0100.*h5  to  MOD04_L2.A2017090.0155.*h5 at every 5 min interval, given that some files are missing:

bash -c "ls MOD04_L2.A2017090.0145.*h5"
ls: MOD04_L2.A2017090.0145.*h5: No such file or directory

bash -c "ls MOD04_L2.A2017090.0150.*h5"
MOD04_L2.A2017090.0150.051.NRT.h5

The * is needed because the filename itself sometimes contains file-processing time, which cannot be represented by time-template. 

In the source code, it is used in two places wrapped with MPI  irank=0 (mapl_am_I_root()), because otherwise this code crashes as multiple processes will write to the same file. 

- https://github.com/GEOS-ESM/MAPL/blob/feature/ygyu/mapl_obsmodule/base/MAPL_SwathGridFactory.F90#L538-L557

- https://github.com/GEOS-ESM/MAPL/blob/feature/ygyu/mapl_obsmodule/gridcomps/History/MAPL_HistoryTrajectoryMod_smod.F90#L546-L553

Reply:
@tclune @metdyn, no, I mean you are asking if there some sort of utility that simulations an "ls" with a wildcard... no, how would one even do that, you can only inquire/open given a file path, if you don't know the path, that's just not something Fortarn can do, you are asking an impossibility without file system commands.

I would say write a script that makes symlink with the untemplatable part stripped. So then no need to do a system call. That's what I would have done, given untemplable file paths for Fortran is unacceptable in my mind, that's what symlinks and a workflow if for before you run.

Reply:
OK, thank you, @bena-nasa.  I will delete this `ls ` system call in Fortran and  prevent a potential flaw in the code.  Thank you for the discussion.

Reply:
The more I think about, this "ls" needs go, maybe fine for development, but the fact is depending what's in the working directory when you do an "ls", it can be very slow. If you ever tried doing an ls in the temporary scratch that exists during DAS run you know what I mean. I'm just worried that this could be very bad for performance too depending on the circumstances.

So now I'm even more convinced this needs to be something addressed before execution as part of the workflow (i.e. make symlinks that have stripped the arbitrary part of the file names)